velox/functions/prestosql/ArrayPosition.cpp

```diff
@@ -0,0 +1,339 @@
+        flatResult.set(row, i + 1);
+        return;
+      }
+    }
+    flatResult.set(row, 0);
+  });
+}
+
+void applyComplexTypeFirstMatch(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = arrayDecoded.indices();
+
+  auto elementsBase = elementsDecoded.base();
+
+  auto searchBase = searchDecoded.base();
+  auto searchIndices = searchDecoded.indices();
+
+  rows.applyToSelected([&](auto row) {
+    auto size = rawSizes[indices[row]];
+    auto offset = rawOffsets[indices[row]];
+    auto searchIndex = searchIndices[row];
+
+    for (auto i = 0; i < size; i++) {
+      if (!elementsBase->isNullAt(offset + i) &&
+          elementsBase->equalValueAt(searchBase, offset + i, searchIndex)) {
+        flatResult.set(row, i + 1);
+        return;
+      }
+    }
+    flatResult.set(row, 0);
+  });
+}
+
+template <>
+void applyTypedFirstMatch<TypeKind::ARRAY>(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  applyComplexTypeFirstMatch(
+      rows, arrayDecoded, elementsDecoded, searchDecoded, flatResult);
+}
+
+template <>
+void applyTypedFirstMatch<TypeKind::MAP>(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  applyComplexTypeFirstMatch(
+      rows, arrayDecoded, elementsDecoded, searchDecoded, flatResult);
+}
+
+template <>
+void applyTypedFirstMatch<TypeKind::ROW>(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    FlatVector<int64_t>& flatResult) {
+  applyComplexTypeFirstMatch(
+      rows, arrayDecoded, elementsDecoded, searchDecoded, flatResult);
+}
+
+template <TypeKind kind>
+void applyTypedWithInstance(
+    const SelectivityVector& rows,
+    DecodedVector& arrayDecoded,
+    const DecodedVector& elementsDecoded,
+    DecodedVector& searchDecoded,
+    const DecodedVector& instanceDecoded,
+    FlatVector<int64_t>& flatResult) {
+  using T = typename TypeTraits<kind>::NativeType;
+
+  auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+  auto rawSizes = baseArray->rawSizes();
+  auto rawOffsets = baseArray->rawOffsets();
+  auto indices = arrayDecoded.indices();
+
+  rows.applyToSelected([&](auto row) {
+    auto size = rawSizes[indices[row]];
+    auto offset = rawOffsets[indices[row]];
+    auto search = searchDecoded.valueAt<T>(row);
+
+    auto instance = instanceDecoded.valueAt<int64_t>(row);
```

Comment:
I expect most cases would have constant "instance" and wonder if it makes sense to provide a fast path for these, e.g. process instance value outside of the loop. It would be nice to add a benchmark to compare performance of this function with hand-written loop for a flat input of primitive types. Here is an example: https://github.com/facebookincubator/velox/pull/492

Another common case could be constant value for "search" parameter.

Reply:
Thank you for this suggestion! I'll add a benchmark in a following PR.